torch/distributed/_shard/sharding_spec/api.py

```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+        split_size = get_split_size(sharding_dim_size, chunks)
+        scatter_shape = list(tensor.size())
+        scatter_shape[self.dim] = split_size  # type: ignore[index]
 
         for shard_meta in tensor_meta.shards_metadata:
             rank, device = _parse_and_validate_remote_device(process_group, shard_meta.placement)
-            if rank == current_rank:
+            if current_rank == src_rank:
                 # Reshape to get shard for this rank and we don't want autograd
                 # recording here for the narrow op and 'local_shard' should be a
                 # leaf variable in the autograd graph.
-                local_tensor = narrow_tensor(tensor, shard_meta).clone().detach().contiguous()
-
-                # Sync requires_grad to local_shard.
-                local_tensor.requires_grad = tensor.requires_grad
-                local_shards.append(
-                    Shard(
-                        tensor=local_tensor,
-                        metadata=shard_meta
-                    )
-                )
+                narrowed_tensor = narrow_tensor(tensor, shard_meta)
+                if shard_meta.shard_sizes[self.dim] < split_size:  # type: ignore[index]
+                    # for the last shard that might be smaller to other shards
+                    # resize the narrowed tensor to the same size and use it for
+                    # the scatter collective as dist.scatter requires same size
+                    # inputs on every rank
+                    tensor_to_scatter = narrowed_tensor.detach().clone().resize_(scatter_shape)
+                else:
+                    tensor_to_scatter = narrowed_tensor.detach().clone().contiguous()
+
+                tensors_to_scatter[rank] = tensor_to_scatter
+
+            if current_rank == rank:
+                local_tensor = torch.empty(
+                    scatter_shape, dtype=tensor.dtype, layout=tensor.layout, device=device)
+                local_metadata = shard_meta
+
+        # each rank should have local_tensor and local_metadata initialized if we build
+        # the metadata list in a correct way.
+        assert local_tensor is not None
+        assert local_metadata is not None
+
+        # Scatter the shards to all ranks in the pg
+        dist.scatter(
+            local_tensor,
+            scatter_list=tensors_to_scatter if current_rank == src_rank else None,
+            src=src_rank,
+            group=process_group
+        )
+
+        if list(local_tensor.size()) != local_metadata.shard_sizes:
+            # detach again after receiving to ensure local shards remain a leaf node
+            local_tensor = local_tensor.resize_(local_metadata.shard_sizes).detach()
+
+        # Sync requires_grad to local_shard.
+        local_tensor.requires_grad = tensor.requires_grad
+
+        local_shards.append(Shard(tensor=local_tensor, metadata=local_metadata))
+
+        st = ShardedTensor._init_from_local_shards_and_global_metadata(
```

Comment:
This will also fix the metadata order we have discussed about before, right? I mean when the placements are like: 
```
placements=[
    "rank:2/cuda:2",
    "rank:3/cuda:3",
    "rank:0/cuda:0",
    "rank:1/cuda:1",
]
```

Reply:
Yep it should fix it